src/main/java/ch/epfl/biop/imaris/EasyXT.java

```diff
@@ -1285,45 +1288,47 @@ public static ISurfacesPrx createFromLabels(ImagePlus impLabel) throws Error {
 
             // build empty surface object
             ISurfacesPrx surface = EasyXT.Utils.getImarisApp().GetFactory().CreateSurfaces();
 
-            //TODO fix issues with time-lapse Label
-            // not all labels per time point
             for (int t = 0; t < impLabel.getNFrames(); t++) {
                 // get the current t stack
                 ImagePlus tImpLabel = new Duplicator().run(impLabel, 1, 1, 1, impLabel.getNSlices(), t + 1, t + 1);
                 // get the minimum and Maximum value of the current t, Labels
                 //int impMin = (int) new StackStatistics(tImpLabel).min; // will always return 0 !
                 int impMax = (int) new StackStatistics(tImpLabel).max;
 
-                // TODO optimize idx start value, see below
-                for (int idx = 1; idx <= impMax; idx++) {
-
-                    ImagePlus tempImage = tImpLabel.duplicate();
-                    IJ.setThreshold(tempImage, idx, idx);
-                    IJ.run(tempImage, "Convert to Mask", "method=Default background=Dark black");
-
-                    int nProcessor = tempImage.getStack().getSize();
-                    IntStream.range(0, nProcessor).parallel().forEach(index -> {
-                        tempImage.getStack().getProcessor(index + 1).multiply(1.0 / 255.0);
-                    });
-
-                    // here we check if there are pixels at 1 , eg pixels have been thresholded
-                    // to avoid trying to create an Empty Surface that is causing an Imaris error
-                    // TODO find a better way to do it, possibly by optimizing the idx start
-                    //  maybe using the histogram ? issue with 16-bit images  (histogram has only 256 bins)?
-                    //  open to suggestions...
-                    int tImpMax = (int) new StackStatistics(tempImage).max;
-                    if (tImpMax == 1) {
+                // we use findAllLabels() and voxelCount() from MorhopholibJ to "simplify" Labels processing
+                int[] labels = LabelImages.findAllLabels(tImpLabel);
+                int[] voxelCounts = LabelImages.voxelCount(tImpLabel.getStack(), labels);
+
+                for (int idx = 1; idx < labels.length; idx++) {
+                    if (voxelCounts[idx] > 1 ) {
+                        // duplicate and theshold a Label
+                        ImagePlus tempImage = tImpLabel.duplicate();
+                        IJ.setThreshold(tempImage, labels[idx], labels[idx]);
+                        IJ.run(tempImage, "Convert to Mask", "method=Default background=Dark black");// can't leave options blank, GUI pops-up
+
+                        // imaris requires binary 0-1
+                        int nProcessor = tempImage.getStack().getSize();
+                        IntStream.range(0, nProcessor).parallel().forEach(index -> {
+                            tempImage.getStack().getProcessor(index + 1).multiply(1.0 / 255.0);
+                        });
```

Comment:
Don't you have an equivalent of all this directly in MorphoLibJ ? Something like LabelImages.getMask(label) ?

Reply:
yes and no ... there is keepLabels but does not simplify so much...